apps/src/dance/blockly/buildDanceBlockly.ts

```diff
@@ -534,43 +542,71 @@ export default function buildDanceBlockly(
   };
 
   // Event blocks for each measure
-  const eventBlocks: BlockState[] = measures
-    // First measure is redundant with "when run"
-    .filter(measure => measure !== 1)
-    .map((measure): BlockState => {
-      const backgroundBlock = makeSetBackgroundBlock(
-        setBackgroundBlockType,
-        validBackgrounds,
-        validPalettes,
-        simpleCode
-      );
-      const foregroundBlock = makeSetForegroundBlock(
-        setForegroundBlockType,
-        validForegrounds,
-        simpleCode
-      );
-      const changeMoveBlock = makeChangeMoveEachLRBlock(
-        changeMoveBlockType,
-        validMoves,
-        backgroundDancers,
-        simpleCode
-      );
-      const chain = chainBlocks(
-        backgroundBlock,
-        foregroundBlock,
-        changeMoveBlock
-      );
-
-      return {
-        ...eventBlock,
-        fields: {
-          TIMESTAMP: measure,
-          UNIT: unitMeasuresField(),
-        },
-        next: {block: chain},
-        kind: 'block',
-      };
+  const eventBlocks: BlockState[] = [];
+  let lastEventBackground: string | null = initialBackgroundEffect;
+  let lastEventForeground: string | null = initialForegroundEffect;
+  let lastEventMove: string | null = initialMoveValue;
+
+  for (const measure of measures.filter(m => m !== 1)) {
+    // Exclude the previously used background effect
+    const backgroundBlock = makeSetBackgroundBlock(
+      setBackgroundBlockType,
+      validBackgrounds,
+      validPalettes,
+      simpleCode,
+      lastEventBackground
+    );
+
+    // Exclude the previously used foreground effect
+    const foregroundBlock = makeSetForegroundBlock(
+      setForegroundBlockType,
+      validForegrounds,
+      simpleCode,
+      lastEventForeground
+    );
+
+    // Exclude the previously used move
+    const changeMoveBlock = makeChangeMoveEachLRBlock(
+      changeMoveBlockType,
+      validMoves,
+      backgroundDancers,
+      simpleCode,
+      undefined, // defaultGroupFieldValue, only used for simple code
+      lastEventMove
+    );
+
+    const chain = chainBlocks(
+      backgroundBlock,
+      foregroundBlock,
+      changeMoveBlock
+    );
+
+    // Capture the actual chosen values from the generated field XML,
+    // so they are not repeated in the next stack.
+    const chosenBg =
+      Blockly.Xml.textToDom((backgroundBlock.fields?.EFFECT as string) || '')
+        .textContent || null;
+    const chosenFg =
+      Blockly.Xml.textToDom((foregroundBlock.fields?.EFFECT as string) || '')
+        .textContent || null;
+    const chosenMove =
+      Blockly.Xml.textToDom((changeMoveBlock.fields?.MOVE as string) || '')
+        .textContent || null;
```

Comment:
Worth extracting into a helper function since it looks very similar to the code above for iniital values?